src/audio/chain_dma.c

```diff
@@ -37,56 +48,478 @@ static struct comp_dev *chain_dma_create(const struct comp_driver *drv,
+
+	if (cd->chain_task.state == SOF_TASK_STATE_FREE)
+		return IPC4_SUCCESS;
+
+	key = k_spin_lock(&drivers->lock);
+	cd->first_data_received = false;
+
+	if (chain_stop(dev) < 0) {
+		k_spin_unlock(&drivers->lock, key);
+		return IPC4_FAILURE;
+	}
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	k_spin_unlock(&drivers->lock, key);
+	schedule_task_free(&cd->chain_task);
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+
 	return IPC4_SUCCESS;
 }
 
-int chain_dma_trigger(struct comp_dev *dev, int cmd)
+int chain_task_trigger(struct comp_dev *dev, int cmd)
 {
 	switch (cmd) {
 	case COMP_TRIGGER_START:
-		return chain_dma_start(dev);
+		return chain_task_start(dev);
 	case COMP_TRIGGER_PAUSE:
-		return chain_dma_pause(dev);
+		return chain_task_pause(dev);
 	default:
 		return IPC4_ERROR_INVALID_PARAM;
 	}
 }
 
-void chain_dma_remove(struct comp_dev *dev)
+void chain_task_free(struct comp_dev *dev)
 {
 	struct chain_dma_data *cd = comp_get_drvdata(dev);
 
+	chain_release(dev);
 	rfree(cd);
 	rfree(dev);
 }
 
+static enum task_state chain_task_run(void *data)
+{
+	struct dma_status stat;
+	size_t link_avail_bytes, link_free_bytes, host_avail_bytes, host_free_bytes;
+	uint32_t link_read_pos, host_read_pos;
+	struct chain_dma_data *cd = (struct chain_dma_data *)data;
+
+	/* retrieve current status of link DMA */
+	int ret = dma_get_status(cd->chan_link->dma->z_dev, cd->chan_link->index,
+				 &stat);
+	if (ret < 0)
+		tr_err(&chain_dma_tr, "chain_task_run(): dma_get_status() failed, ret = %u",
+			 ret);
+
+	link_avail_bytes = stat.pending_length;
+	link_free_bytes = stat.free;
+	link_read_pos = stat.read_position;
+
+
+	/* retrieve current status of host DMA */
+	ret = dma_get_status(cd->chan_host->dma->z_dev, cd->chan_host->index, &stat);
+	if (ret < 0)
+		tr_err(&chain_dma_tr, "chain_task_run(): dma_get_status() failed, ret = %u",
+		       ret);
+
+	host_avail_bytes = stat.pending_length;
+	host_free_bytes = stat.free;
+	host_read_pos = stat.read_position;
+
+	uint32_t link_type = cd->link_connector_node_id.f.dma_type;
+
+	if (link_type == ipc4_hda_link_input_class) {
+		const size_t increment = MIN(host_free_bytes, link_avail_bytes);
+
+		ret = dma_reload(cd->chan_host->dma->z_dev, cd->chan_host->index,
+				 0, 0, increment);
+		if (ret < 0)
+			tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() host failed, ret = %u",
+			       ret);
+
+		ret = dma_reload(cd->chan_link->dma->z_dev, cd->chan_link->index, 0, 0,
+				 increment);
+		if (ret < 0)
+			tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() link failed, ret = %u",
+			       ret);
+	} else {
+		const size_t half_buff_size = cd->dma_buffer->stream.size / 2;
+
+		if ((!cd->first_data_received) && (host_avail_bytes > half_buff_size)) {
+			ret = dma_reload(cd->chan_link->dma->z_dev,
+					 cd->chan_link->index, 0, 0,
+					 half_buff_size);
+			if (ret < 0)
+				tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() link failed, ret = %u",
+				       ret);
```

Comment:
make these lines shorter by breaking it down.
and is this error or warn?

Reply:
These all looks more like warning, so I change it accordingly.